Add client-side routing section to cluster routing page (#867)
neo4j/docs-operations: modules/ROOT/pages/clustering/setup/routing.adoc

By default, Neo4j avoids this by automatically transferring database leaderships so that they are evenly distributed throughout the cluster.
Additionally, Neo4j automatically transfers database leaderships away from instances where those databases are configured to be read-only using xref:configuration/configuration-settings.adoc#config_server.databases.read_only[server.databases.read_only] or similar.

[[clustering-client-side-routing]]
== Client-side routing

Client-side routing is when the database client takes control over which cluster member to send specific requests to.

> Typically this would be to make sure that write operations are sent to the server that can write for your specific database, and that read operations are sent to other servers.

Reviewer: ```suggestion
Typically this would be to make sure that write operations are sent to the server that can write for the target database, and that read operations are sent to other servers.
```

> Typically this would be to make sure that write operations are sent to the server that can write for your specific database, and that read operations are sent to other servers.

Author: The column in `SHOW SERVERS` is called `writer`. Leader only applies to Raft groups, and is not really a user-facing concept. So for instance databases with a single primary do not in fact have a leader, though they do have a writer.

> Typically this would be to make sure that write operations are sent to the server that can write for your specific database, and that read operations are sent to other servers.

Reviewer: Ah great to know, thanks for the clarification! :)

Client-side routing is based on getting a routing table from a cluster member, and then using that information to make the routing decisions.

> A routing table provides the writers, readers and routers for a specific database.

Reviewer: ```suggestion
A routing table provides the writers, readers, and routers with routing information for a specific database.
```

Author: Superseded by other suggestion.

> A routing table provides the writers, readers and routers for a specific database.

Reviewer: ```suggestion
A routing table contains information about the writers, readers, and routers for a specific database.
```

`Provide` sounds ambiguous to me, as I kept looking for a `with` predicate that I couldn't find (as in "the routing table provides me with joy")

> There will usually be one writer, though there may be none if the database is read only or unhealthy.

Reviewer: ```suggestion
There is usually one _writer_, though there may be none if the database is read only or unhealthy.
```

> With the default configuration all other servers that host the database will be considered readers, i.e. the writer will not be in the list of readers.

Reviewer: ```suggestion
With the default configuration, all other servers that host the database are considered _readers_, i.e. the writer is not in the list of readers.
```

Author: I have got to cut down on the future tense...

> This is to let it focus on the write load and not have to manage two kinds of interaction.

Reviewer: ```suggestion
This is to let it focus on the write load and not have to manage two kinds of interactions.
```

> Typically all members of the cluster will be given as 'routers', which are servers that can be contacted to get a new routing table.

Reviewer 2: Have you checked that? Are routers not only members whihc actually host the db?
I agree it could be everyone, but I don't we chnaged this.

Author: You are quite right, I thought we had changed that, but it's just the hosts. Will update, thanks.

> Typically all servers that host the database will be given as 'routers', which are servers that can be contacted to get a new routing table for that database.

Reviewer: ```suggestion
Typically, all servers that host the database are listed as _routers_, which are servers that can be contacted to get a new routing table for that database.
```

> Our standard drivers will get a routing table the first time they attempt to connect to a database, and will then get a fresh one after the configured time-to-live, or if it seems the routing table has become out of date.

Reviewer: ```suggestion
link:{neo4j-docs-base-uri}/create-applications/[Neo4j Drivers] retrieve a routing table the first time they attempt to connect to a database, and fetch a fresh one after the configured time-to-live, or if it seems the routing table has got out of date.
```

> For example, if the routing table says `server-3` is the writer for the database, but write requests get rejected with 'not able to write', the driver may decide to get a new routing table, because the writer could be a different server.

Reviewer: ```suggestion
For example, if the routing table lists `server-3` as the writer for the database, but write requests get rejected with a _not able to write_ error, the driver may decide to get a new routing table, because the writer could be a different server.
```

[[clustering-client-side-routing-policies]]
=== Routing policies

You can control the routing table that servers provide by using xref:clustering/clustering-advanced/multi-data-center-routing.adoc[routing policies].
Policies filter the full set of possible servers for each category according to the rules you define.

> For example this can be used to preferentially route to a local data centre, or to specific large machines, depending on your policies.

Reviewer: ```suggestion
For example, this can be used to preferentially route to a local data centre, or to specific large machines, depending on your policies.
```

[[clustering-client-side-routing-table-api]]
=== Routing table API

> If you want lower level detail about getting routing tables, you can read the link:{neo4j-docs-base-uri}/bolt/current/driver-api/#_driver_routing_table[Bolt driver API].

Reviewer: ```suggestion
For lower level details about getting routing tables, refer to the link:{neo4j-docs-base-uri}/bolt/current/driver-api/#_driver_routing_table[Bolt protocol documentation].
```

> If you want lower level detail about getting routing tables, you can read the link:{neo4j-docs-base-uri}/bolt/current/driver-api/#_driver_routing_table[Bolt driver API].

Reviewer: A single sentence is very short for a separate section. I'd recommend moving this as concluding remark in the client-side routing section, as a separate paragraph :)
